File: app/services/activity.py
from __future__ import annotations

import logging

from sqlalchemy.orm import Session

from app.models import ActivityLog

logger = logging.getLogger("indexmeplease.activity")


def log_event(
    db: Session,
    message: str,
    *,
    workspace_id: int | None = None,
    level: str = "info",
    category: str = "system",
    details: dict | None = None,
    commit: bool = False,
) -> ActivityLog:
    entry = ActivityLog(
        workspace_id=workspace_id,
        level=level,
        category=category,
        message=message,
        details=details,
    )
    db.add(entry)
    if commit:
        db.commit()
    else:
        db.flush()
    logger.log(logging.WARNING if level in ("warning", "error") else logging.INFO, message)
    return entry

File: src/services/activity.ts
import { db } from "../db";
import type { ActivityEntry } from "../db/types";

export interface LogOptions {
  workspaceId?: number | null;
  level?: "info" | "warning" | "error";
  category?: string;
  details?: unknown;
}

export async function logEvent(message: string, options: LogOptions = {}): Promise<void> {
  const { workspaceId = null, level = "info", category = "system", details } = options;

  await db
    .insertInto("activity_log")
    .values({
      workspace_id: workspaceId,
      level,
      category,
      message,
      // MySQL oczekuje tekstu w kolumnie JSON, nie obiektu.
      details: details === undefined ? null : JSON.stringify(details),
    })
    .execute();

  const line = `[${category}] ${message}`;
  if (level === "warning" || level === "error") console.warn(line);
  else console.log(line);
}

export function recentActivity(workspaceId: number, limit = 50): Promise<ActivityEntry[]> {
  return db
    .selectFrom("activity_log")
    .selectAll()
    .where((eb) =>
      eb.or([eb("workspace_id", "=", workspaceId), eb("workspace_id", "is", null)]),
    )
    .orderBy("created_at", "desc")
    .limit(limit)
    .execute();
}
